File: libs/monobitmap/src/main/kotlin/mono/graphics/bitmap/drawable/NinePatchDrawable.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.graphics.bitmap.drawable

import mono.common.Characters.TRANSPARENT_CHAR
import mono.graphics.bitmap.MonoBitmap
import kotlin.math.max
import kotlin.math.min

/**
 * A simple 9-patch image which scales image based on repeating points in
 * [horizontalRepeatableRange] and [horizontalRepeatableRange].
 */
class NinePatchDrawable(
    private val pattern: Pattern,
    private val horizontalRepeatableRange: RepeatableRange =
        RepeatableRange.Scale(0, pattern.width - 1),
    private val verticalRepeatableRange: RepeatableRange =
        RepeatableRange.Scale(0, pattern.height - 1)
) : Drawable {
    override fun toBitmap(width: Int, height: Int): MonoBitmap {
        val builder = MonoBitmap.Builder(width, height)
        val rowIndexes = verticalRepeatableRange.toIndexes(height, pattern.height)
        val colIndexes = horizontalRepeatableRange.toIndexes(width, pattern.width)
        for (row in 0 until height) {
            for (col in 0 until width) {
                builder.put(row, col, pattern.getChar(rowIndexes[row], colIndexes[col]))
            }
        }
        return builder.toBitmap()
    }

    /**
     * A data class which provides a basic render characters for 9-patch image.
     */
    data class Pattern(val width: Int, val height: Int, private val chars: List<Char>) {
        init {
            if (chars.size < width * height) {
                throw IllegalArgumentException("Mismatch between size and number of chars provided")
            }
        }

        internal fun getChar(row: Int, column: Int): Char {
            val index = row * width + column
            return chars[index]
        }

        companion object {
            fun fromText(
                text: String,
                delimiter: Char = '\n',
                transparentChar: Char = ' '
            ): Pattern {
                val array = text.split(delimiter)
                if (array.isEmpty()) {
                    return Pattern(0, 0, emptyList())
                }
                val width = array.first().length
                val height = array.size
                val chars = array.joinToString("")
                    .map { if (it == transparentChar) TRANSPARENT_CHAR else it }
                return Pattern(width, height, chars)
            }
        }
    }

    /**
     * The algorithm for repeating the repeated range.
     */
    sealed class RepeatableRange(
        start: Int,
        endInclusive: Int
    ) {
        private val start: Int = max(0, min(start, endInclusive))
        private val endInclusive: Int = max(start, endInclusive)

        /**
         * Creates a list with size [size] of indexes whole value in range [0, [patternSize]).
         * If [patternSize] < [endInclusive], [endInclusive] will be used for the index value range.
         */
        fun toIndexes(size: Int, patternSize: Int): List<Int> {
            val adjustedEndInclusive = min(patternSize - 1, endInclusive)
            val rangeSize = adjustedEndInclusive - start + 1
            val repeatingLeft = start
            val repeatingRight = size - (patternSize - adjustedEndInclusive)

            return List(size) { index ->
                when {
                    index < repeatingLeft -> index
                    index > repeatingRight -> adjustedEndInclusive + index - repeatingRight
                    else -> scaleIndex(index, repeatingLeft, repeatingRight, rangeSize)
                }
            }
        }

        protected abstract fun scaleIndex(
            index: Int,
            minRepeatingIndex: Int,
            maxRepeatingIndex: Int,
            rangeSize: Int
        ): Int

        /**
         * The algorithm for repeating the repeated range: `01` -> `00001111`
         */
        class Scale(start: Int, endInclusive: Int) : RepeatableRange(start, endInclusive) {

            override fun scaleIndex(
                index: Int,
                minRepeatingIndex: Int,
                maxRepeatingIndex: Int,
                rangeSize: Int
            ): Int {
                val repeatingSize = maxRepeatingIndex - minRepeatingIndex + 1
                return minRepeatingIndex + (index - minRepeatingIndex) * rangeSize / repeatingSize
            }
        }

        /**
         * The algorithm for repeating the repeated range: `01` -> `01010101`
         */
        class Repeat(start: Int, endInclusive: Int) : RepeatableRange(start, endInclusive) {

            override fun scaleIndex(
                index: Int,
                minRepeatingIndex: Int,
                maxRepeatingIndex: Int,
                rangeSize: Int
            ): Int = minRepeatingIndex + (index - minRepeatingIndex) % rangeSize
        }
    }
}

File: monosketch-svelte/src/lib/mono/monobitmap/drawable/nine-patch-drawable.ts
import type { Char } from "$libs/char";
import { TRANSPARENT_CHAR } from "$mono/common/character";
import { MonoBitmap } from "$mono/monobitmap/bitmap/monobitmap";
import type { Drawable } from "$mono/monobitmap/drawable/drawable";

/**
 * A simple 9-patch image which scales image based on repeating points in
 * horizontal and vertical repeatable range.
 */
export class NinePatchDrawable implements Drawable {
    private pattern: NinePatchDrawablePattern;
    private horizontalRepeatableRange: NinePatchDrawableRepeatableRange;
    private verticalRepeatableRange: NinePatchDrawableRepeatableRange;

    constructor(
        pattern: NinePatchDrawablePattern,
        horizontalRepeatableRange: NinePatchDrawableRepeatableRange = new ScaleRepeatableRange(0, pattern.width - 1),
        verticalRepeatableRange: NinePatchDrawableRepeatableRange = new ScaleRepeatableRange(0, pattern.height - 1),
    ) {
        this.pattern = pattern;
        this.horizontalRepeatableRange = horizontalRepeatableRange;
        this.verticalRepeatableRange = verticalRepeatableRange;
    }

    toBitmap(width: number, height: number): MonoBitmap.Bitmap {
        const builder = new MonoBitmap.Builder(width, height);
        const rowIndexes = this.verticalRepeatableRange.toIndexes(height, this.pattern.height);
        const colIndexes = this.horizontalRepeatableRange.toIndexes(width, this.pattern.width);
        for (let row = 0; row < height; row++) {
            for (let col = 0; col < width; col++) {
                builder.put(
                    row,
                    col,
                    this.pattern.getChar(rowIndexes[row], colIndexes[col]),
                    // TODO: Think about a way to support direction chars to 9-patch drawable
                    this.pattern.getChar(rowIndexes[row], colIndexes[col]),
                );
            }
        }
        return builder.toBitmap();
    }
}

/**
 * A data class which provides a basic render characters for 9-patch image.
 */
export class NinePatchDrawablePattern {

    constructor(public width: number, public height: number, private chars: Char[]) {
        if (chars.length < width * height) {
            throw new Error("Mismatch between size and number of chars provided");
        }
    }

    getChar(row: number, column: number): Char {
        const index = row * this.width + column;
        return this.chars[index];
    }

    static fromText(text: string, delimiter: Char = '\n', transparentChar: Char = ' '): NinePatchDrawablePattern {
        const array = text.split(delimiter);
        if (array.length === 0) {
            return new NinePatchDrawablePattern(0, 0, []);
        }
        const width = array[0].length;
        const height = array.length;
        const chars = array.map((char) => char === transparentChar ? TRANSPARENT_CHAR : char);
        return new NinePatchDrawablePattern(width, height, chars);
    }
}

/**
 * The algorithm for repeating the repeated range.
 */
export abstract class NinePatchDrawableRepeatableRange {
    protected start: number;
    protected endInclusive: number;

    constructor(start: number, endInclusive: number) {
        this.start = Math.max(0, Math.min(start, endInclusive));
        this.endInclusive = Math.max(start, endInclusive);
    }

    /**
     * Creates a list with size [size] of indexes whole value in range [0, [patternSize]).
     * If [patternSize] < [endInclusive], [endInclusive] will be used for the index value range.
     */
    toIndexes(size: number, patternSize: number): number[] {
        const adjustedEndInclusive = Math.min(patternSize - 1, this.endInclusive);
        const rangeSize = adjustedEndInclusive - this.start + 1;
        const repeatingLeft = this.start;
        const repeatingRight = size - (patternSize - adjustedEndInclusive);

        return Array.from({ length: size }, (_, index) => {
            if (index < repeatingLeft) return index;
            if (index > repeatingRight) return adjustedEndInclusive + index - repeatingRight;
            return this.scaleIndex(index, repeatingLeft, repeatingRight, rangeSize);
        });
    }

    protected abstract scaleIndex(index: number, minRepeatingIndex: number, maxRepeatingIndex: number, rangeSize: number): number;
}

/**
 * The algorithm for repeating the repeated range: `01` -> `00001111`
 */
export class ScaleRepeatableRange extends NinePatchDrawableRepeatableRange {
    scaleIndex(index: number, minRepeatingIndex: number, maxRepeatingIndex: number, rangeSize: number): number {
        const repeatingSize = maxRepeatingIndex - minRepeatingIndex + 1;
        return minRepeatingIndex + Math.floor((index - minRepeatingIndex) * rangeSize / repeatingSize);
    }
}

/**
 * The algorithm for repeating the repeated range: `01` -> `01010101`
 */
export class RepeatRepeatableRange extends NinePatchDrawableRepeatableRange {
    scaleIndex(index: number, minRepeatingIndex: number, maxRepeatingIndex: number, rangeSize: number): number {
        return minRepeatingIndex + (index - minRepeatingIndex) % rangeSize;
    }
}